### ISSUE TITLE
Gradle jars "grossly invalid zip files" that trigger zip bomb detection [CVE-2019-13232]

### ISSUE THREAD

**Issue by bencaradocdavies** [gradle/gradle#9990]
Gradle creates jar files that are "grossly invalid zip files", triggering zip bomb detection introduced into unzip by Mark Adler to address [CVE-2019-13232](https://nvd.nist.gov/vuln/detail/CVE-2019-13232). Several affected files are also included in the Gradle 5.5.1 binary distribution.

Impact is that affected jar files cannot be unzipped by unzip patched to address CVE-2019-13232. Furthermore, duplication and orphaned chunks waste space and reduce efficiency of compression. Expected outcome is that jar files are valid zip files, can be unzipped, and do not contain wasteful duplicates or orphaned chunks.

[Mark Adler writes](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=931895#73):
"In summary, those jar files are grossly invalid zip files for three reasons: 1. The same name appears twice, 2. Two central directory entries point to the same local header (setting off zip bomb detection), and 3. There is an orphaned chunk of the file that is not referred to by any central directory header."

17 affected files from ASM, Spring, and Gradle (including 5.5.1) are contained in four zip files attached to [Debian Bug#931895](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=931895). Affected jar files included in the Gradle 5.5.1 binary distribution are:
gradle-kotlin-dsl-5.5.1.jar
gradle-kotlin-dsl-provider-plugins-5.5.1.jar
gradle-kotlin-dsl-tooling-builders-5.5.1.jar
groovy-all-1.0-2.5.4.jar

See also [Debian Bug#931433](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=931433).

**Comment by bencaradocdavies** [comment]
And you can see the erroneous duplication even with an unpatched unzip by unzipping one of the affected files in an empty directory ([as noticed by Mark Adler](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=931895#73)) :
```$ unzip ../gradle-kotlin-dsl-5.5.1.jar 
Archive:  ../gradle-kotlin-dsl-5.5.1.jar
[...]
   creating: META-INF/services/
  inflating: META-INF/services/org.gradle.initialization.GradleApiSpecProvider  
  inflating: META-INF/services/org.gradle.internal.service.scopes.PluginServiceRegistry  
  inflating: gradle-kotlin-dsl-classpath.properties  
replace META-INF/services/org.gradle.initialization.GradleApiSpecProvider? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
  inflating: META-INF/services/org.gradle.initialization.GradleApiSpecProvider  
replace META-INF/services/org.gradle.internal.service.scopes.PluginServiceRegistry? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
  inflating: META-INF/services/org.gradle.internal.service.scopes.PluginServiceRegistry  
[...]```

There should not be multiple instances of these files. Other affected jar files have duplicate class files so it is not just SPI files that are affected.

Patched unzip fails on this jar file with: `error: invalid zip file with overlapped components (possible zip bomb)`.

**Comment by eskatos** [comment]
Thanks for the report @bencaradocdavies !

This might be due to https://github.com/gradle/gradle/issues/10005, we are still investigating.

Gradle 5.6 will already have the `gradle-kotlin-dsl-*.jar` archives fixed to not contain duplicated entries, see https://github.com/gradle/gradle/pull/10002

**Comment by lptr** [comment]
IIUC we've fixed `groovy-all.zip` as well in 5.6 (and added tests for all the ZIPs shipped/generated by Gradle). This should be good to close. WDYT @paplorinc?

**Comment by paplorinc** [comment]
Yes, this is fixed in 5.6-RC1.
Thanks @bencaradocdavies for the issue, closing it. If you still find errors, please reopen it.
